### ISSUE TITLE
Replace hidden glyph visualization code

### ISSUE THREAD

**Issue by shysaur** [shysaur/Fragaria#44]
I feel like the way hidden glyphs are made visible is currently a bit hacky. But, from Apple's documentation, how one is supposed to implement this feature in the right way is not clear. Plus, none of Apple's apps which use Cocoa's text system implement this feature (iWork has its own CoreText based text system), which is telling of what Apple thinks about it... 

Currently, at the end of the layout process, SMLLayoutManager searches for invisible characters and draws (with CoreText) a visible character overlapping them.

I'm going to remove the (already disabled) code in SMLLayoutManager and MGSGlyphGenerator which tries to implement this feature by glyph substitution at the glyph generation phase because it's not going to work ever.

The text system relies on having NSGlyphGenerator produce special glyphs for line breaks, tabs and spaces, and replacing them with any other character will break the layout phase (tabs will no longer tab, line breaks will no longer break lines, and I even managed to lock NSATSTypesetter in an infinite loop searching for hypenation points by trying to make this approach work).

I tried to reimplement hidden glyph visualization by subclassing NSATSTypesetter and replacing control glyphs with visible glyphs during the layout phase, which is how hyphens are added when automatic hyphenation is enabled. Basically, I let the typesetter do its work (laying out unmodified glyphs), then each invisible control glyph is replaced with a visible glyph. Since the replacement takes place after the original glyph has been laid out, the replacement glyph gets the same position as the invisible glyph, and no custom drawing is needed.

This works very very well, and requires very very little code (I don't know if it's faster, but it's certainly comparable with the current approach).

But there's an issue: since attributes can't be overridden by the typesetter, the substituted visible glyphs have the same attributes as the surrounding text, which is obviously correct, but we have this feature where invisible glyphs are set a different color, which has some value (TextWrangler has it, for example). This feature can be easily restored by adding a syntax coloring pass which always executes and colors invisible glyphs, but that would probably slow down SMLSyntaxColouring even further.

So, summing up everything, there's no real advantage in the newer code... The new implementation is in the rewrite-invisibles branch (look at MGSTypesetter.m). @balthisar, what do you think?

**Comment by balthisar** [comment]
When I fetch and checkout rewrite-invisibles, I'm seeing two glyphs atop each other in the text viewer:
- The current glyph and behavior, with the designated glyph color (old behavior), and
- The newly implemented behavior; the glyphs are the same as the text color and not the substituted glyphs.

The syntax coloring is fairly quick on my system; adding another pass may or may not slow things down so that they're noticeable. There's the possibility of doing the second pass in a dispatch queue. It would probably keep the text editor responsive, but maybe some possibility of the colors doing the wrong thing if scrolling really quickly (or maybe not, since the actual drawing will be performed on the main queue).

**Comment by shysaur** [comment]
Yeah, I left the old code around for testing/comparison. The new stuff is always enabled in that branch.

**Comment by shysaur** [comment]
Since NSLayoutManager is not thread safe, we can't color anything from a separate thread (if the user edits the text/scrolls the view/does anything with Fragaria while coloring is in progress, things will screw up)

**Comment by shysaur** [comment]
Well, in the end, it was an interesting experiment, but since there are no advantages in respect to the current implementation, it will remain an experiment and the current implementation will stay.
